### PR TITLE
Mark all remaining mutable tables as safe or to-be-removed

### PR DESCRIPTION
**Pull request by snowleopard**
Also delete redundant recalculation of `Ocaml_version`.

### CODE REVIEW

File: src/stdune/interned.ml [hunk 1]
```diff
@@ -70,6 +70,9 @@ module type Settings = sig
 end
 
 module Make (R : Settings) () = struct
+  (* The mutable tables in this module can be made safe if we stop leaking
+     information about the string representation, e.g. by switching from [int]s
+     to an opaque data type. *)
```
Comment (rgrinberg) [inline src/stdune/interned.ml]: Where do we leak information about the string representation? The type that this functor produces is abstract.

Comment (snowleopard) [reply src/stdune/interned.ml]: I update the comment: we expose `compare` that does leak information.

File: src/dune/cached_digest.ml [hunk 2]
```diff
@@ -26,6 +26,11 @@ end)
 
 let needs_dumping = ref false
 
+(* CR-soon amokhov: replace this mutable table with a memoized function. This
+   will probably require splitting this module in two, for dealing with source
+   and target files, respectively. For source files, we receive updates via the
+   file-watching API. For target files, we modify the digests ourselves, without
+   subscribing for file-watching updates. *)
```
Comment (diml) [inline src/dune/cached_digest.ml]: One thing to be mindful about here: memoized functions are currently not persistent, while cached digests for targets are. If we used a memoised function for target digests, I'd expect null builds to get slower.

File: src/dune/dune_project.ml [hunk 3]
```diff
@@ -348,6 +348,8 @@ module Extension = struct
         -> Univ_map.t * Stanza.Parser.t list
     }
 
+  (* CR-soon amokhov: convert this mutable table to a memoized function, which
+     depends on the contents of dune files that declare extensions. *)
```
Comment (diml) [inline src/dune/dune_project.ml]: This is more like the `standard` table for aliases. It's only modified at toplevel.

File: src/dune/findlib/findlib.ml [hunk 4]
```diff
@@ -562,6 +562,11 @@ let all_packages t =
            (Dune_package.Entry.name a)
            (Dune_package.Entry.name b))
 
+(* CR-soon amokhov: Remove the mutable table below and add:
+
+   - A memoized function for finding packages by names (see [find]).
+
+   - A [Memo.Lazy.t] storing the set of all packages (see [root_packages]). *)
```
Comment (diml) [inline src/dune/findlib/findlib.ml]: Seems reasonable.

**Pull request merged**